web/src/settings_users.ts

```diff
@@ -0,0 +1,736 @@
+        section.filters[key] = value;
+    }
+    // This hard_redraw will rerun the relevant predicate function
+    // and in turn apply the new filters.
+    section.list_widget?.hard_redraw();
+}
+
+function role_selected_handler(
+    event: JQuery.ClickEvent,
+    dropdown: tippy.Instance,
+    widget: dropdown_widget.DropdownWidget,
+): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    const role_code = Number($(event.currentTarget).attr("data-unique-id"));
+    if (widget.widget_name === section.active.dropdown_widget_name) {
+        add_value_to_filters(section.active, "role_code", role_code);
+    } else if (widget.widget_name === section.deactivated.dropdown_widget_name) {
+        add_value_to_filters(section.deactivated, "role_code", role_code);
+    }
+
+    dropdown.hide();
+    widget.render();
+}
+
+function get_roles(): {
+    unique_id: number;
+    name: string;
+}[] {
+    return [
+        {unique_id: 0, name: $t({defaultMessage: "All roles"})},
+        ...Object.values(settings_config.user_role_values)
+            .map((user_role_value) => ({
+                unique_id: user_role_value.code,
+                name: user_role_value.description,
+            }))
+            .reverse(),
+    ];
+}
+
+function create_role_filter_dropdown(
+    $events_container: JQuery,
+    section: ActiveSection | DeactivatedSection,
+): void {
+    new dropdown_widget.DropdownWidget({
+        widget_name: section.dropdown_widget_name,
+        unique_id_type: dropdown_widget.DataTypes.NUMBER,
+        get_options: get_roles,
+        $events_container,
+        item_click_callback: role_selected_handler,
+        default_id: section.filters.role_code,
+        tippy_props: {
+            offset: [0, 0],
+        },
+    }).setup();
+}
+
+function populate_users(): void {
+    const active_user_ids = people.get_realm_active_human_user_ids();
+    const deactivated_user_ids = people.get_non_active_human_ids();
+
+    if (active_user_ids.length === 0 && deactivated_user_ids.length === 0) {
+        failed_listing_users();
+    }
+
+    section.active.create_table(active_user_ids);
+    section.deactivated.create_table(deactivated_user_ids);
+    create_role_filter_dropdown($("#admin-user-list"), section.active);
+    create_role_filter_dropdown($("#admin-deactivated-users-list"), section.deactivated);
+}
+
+function reset_scrollbar($sel: JQuery): () => void {
+    return function () {
+        scroll_util.reset_scrollbar($sel);
+    };
+}
+
+function bot_owner_full_name(owner_id: number | null): string | undefined {
+    if (!owner_id) {
+        return undefined;
+    }
+
+    const bot_owner = people.maybe_get_user_by_id(owner_id);
+    if (!bot_owner) {
+        return undefined;
+    }
+
+    return bot_owner.full_name;
+}
+
+function bot_info(bot_user_id: number): BotInfo | undefined {
+    const bot_user = people.maybe_get_user_by_id(bot_user_id);
+
+    if (!bot_user) {
+        return undefined;
+    }
```

Comment:
This is the only reason to have `undefined` in all the `BotInfo | undefined` types. But it’s unreachable: `bot_info` is only called from `update_bot_data`, whose only two callers have both checked `people.is_valid_bot_user` before calling this. So we can remove all those `undefined`s.

```suggestion
    assert(bot_user !== undefined);
```